File: tests/test_attitude_codec.cpp
// Mirrors tests/test_attitude_codec.py — size, round-trip, big-endian layout.
#include <gtest/gtest.h>

#include <cmath>

#include "serial_link/codec.hpp"

using namespace serial_link;
using namespace serial_link::codec;

TEST(AttitudeCodec, BodyIs20Bytes) {
    Bytes info = AttitudeSample{1, 2, 0, 0, 0}.encode();
    EXPECT_EQ(info.size(), 21u);  // 1 type-id + 20 body
    EXPECT_EQ(info[0], 0x20);
}

TEST(AttitudeCodec, RoundtripExactFloats) {
    AttitudeSample msg{1234, 56789, 12.5f, -3.25f, 180.0f};
    Message back = decode(msg.encode());
    ASSERT_TRUE(std::holds_alternative<AttitudeSample>(back));
    const auto& a = std::get<AttitudeSample>(back);
    EXPECT_EQ(a.seq, 1234u);
    EXPECT_EQ(a.t_ms, 56789u);
    EXPECT_EQ(a.roll, 12.5f);
    EXPECT_EQ(a.pitch, -3.25f);
    EXPECT_EQ(a.yaw, 180.0f);
}

TEST(AttitudeCodec, RoundtripInexactWithinF32) {
    AttitudeSample msg{0, 0, 15.0f, -7.123f, 359.9f};
    Message decoded = decode(msg.encode());
    const auto& a = std::get<AttitudeSample>(decoded);
    EXPECT_NEAR(a.pitch, -7.123f, 1e-3);
    EXPECT_NEAR(a.yaw, 359.9f, 1e-3);
}

TEST(AttitudeCodec, BigEndianWireLayout) {
    AttitudeSample msg{1, 2, 1.0f, 2.0f, 3.0f};
    // type_id + >IIfff : seq=1, t_ms=2, roll=1.0, pitch=2.0, yaw=3.0
    Bytes expected = {0x20,
                      0x00, 0x00, 0x00, 0x01,  // seq
                      0x00, 0x00, 0x00, 0x02,  // t_ms
                      0x3F, 0x80, 0x00, 0x00,  // 1.0f
                      0x40, 0x00, 0x00, 0x00,  // 2.0f
                      0x40, 0x40, 0x00, 0x00}; // 3.0f
    EXPECT_EQ(msg.encode(), expected);
}

TEST(AttitudeCodec, EqualityOperator) {
    EXPECT_EQ((AttitudeSample{1, 2, 3, 4, 5}), (AttitudeSample{1, 2, 3, 4, 5}));
    EXPECT_FALSE((AttitudeSample{1, 2, 3, 4, 5}) == (AttitudeSample{1, 2, 3, 4, 6}));
}

File: tools/tests/test_attitude_codec.py
import struct

import pytest

from serial_link import decode, AttitudeSample
from serial_link.codec import registry


def test_type_id_registered():
    assert registry[0x20] is AttitudeSample
    assert AttitudeSample.TYPE_ID == 0x20


def test_body_is_20_bytes():
    info = AttitudeSample(seq=1, t_ms=2, roll=0, pitch=0, yaw=0).encode()
    assert len(info) == 21          # 1 type-id + 20 body
    assert info[0] == 0x20


def test_roundtrip_floats():
    msg = AttitudeSample(seq=1234, t_ms=56789, roll=12.5, pitch=-3.25, yaw=180.0)
    back = decode(msg.encode())
    assert back.seq == 1234 and back.t_ms == 56789
    # 12.5 / -3.25 / 180.0 are exactly representable in float32
    assert (back.roll, back.pitch, back.yaw) == (12.5, -3.25, 180.0)


def test_roundtrip_inexact_floats_within_f32():
    msg = AttitudeSample(seq=0, t_ms=0, roll=15.0, pitch=-7.123, yaw=359.9)
    back = decode(msg.encode())
    assert back.pitch == pytest.approx(-7.123, abs=1e-3)
    assert back.yaw == pytest.approx(359.9, abs=1e-3)


def test_big_endian_wire_layout():
    # type_id + >IIfff matches the documented layout exactly.
    msg = AttitudeSample(seq=1, t_ms=2, roll=1.0, pitch=2.0, yaw=3.0)
    expected = bytes([0x20]) + struct.pack(">IIfff", 1, 2, 1.0, 2.0, 3.0)
    assert msg.encode() == expected
